### PR TITLE
YAML output support

### PR DESCRIPTION
**Pull request by kaniini**
Adds YAML generation support to libifupdown.

At the moment, the generator is minimal and only covers parts of the YAML 1.1 specification directly relevant to ifupdown-ng.

We also add an ifparse(1) utility which is intended to be a configuration-specific query tool that provides raw configuration data in either `pretty-print` or `yaml-raw` format.

The `yaml-raw` format is a type-annotated YAML subset that can be parsed (with annotations) by a conformant YAML 1.1 parser.  It is not intended to be particularly human friendly (indeed, the output format has been optimized for simplest possible YAML generation code), instead the intended use is for other tools to use `yaml-raw` to consume data from the ifupdown-ng config parser in a friendly way.

A "friendly" YAML output format (without type annotations) will be made available at a later date -- this format is strictly intended for "internal" consumption by the other NSL platform components.

### CODE REVIEW

**Review by awilfox — COMMENTED** [review]
This looks mostly fine, but I would suggest two changes for the implementation of `ifparse` itself:

1. Use a -f/--format parameter instead of separate parameters for each format.
    This will make it easier to add new formats later.

2. Use a dispatch table instead of an if tree in the loop in `ifparse_main`.
    If a format isn't enabled in the config, an empty function body that just says "format not available" or such can be used.

File: cmd/ifparse.c [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+		struct lif_dict_entry *entry = iter->data;
+		const char *value = entry->data;
+		char addr_buf[512];
+
+		if (!strcmp(entry->key, "address"))
+		{
+			struct lif_address *addr = entry->data;
+
+			if (!lif_address_unparse(addr, addr_buf, sizeof addr_buf, true))
+				continue;
```
Comment (BarbarossaTM) [inline cmd/ifparse.c:105]: Do we need some kind of error handling / output here?

File: cmd/ifparse.c [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+				prettyprint_interface_yaml(entry->data);
+			else
+#endif
+				prettyprint_interface_eni(entry->data);
+		}
+
+		return EXIT_SUCCESS;
+	}
+
+	if (optind >= argc)
```
Comment (BarbarossaTM) [inline cmd/ifparse.c:173]: Shouldn't this be checked before `show_all`?

**Review by BarbarossaTM — COMMENTED** [review]
+1 to what @awilfox said, plus some minor questions

File: cmd/ifparse.c (hunk 1 above)
Comment (kaniini) [reply cmd/ifparse.c:105]: What did you have in mind?

File: cmd/ifparse.c (hunk 2 above)
Comment (kaniini) [reply cmd/ifparse.c:173]: No, we should keep semantics consistent across all tools, I think.  `--all` means *all*, so we should check for that first and then exit.

File: cmd/ifparse.c (hunk 1 above)
Comment (BarbarossaTM) [reply cmd/ifparse.c:99]: I don't know really. I'm wondering what to do if `lif_address_unprase` fails. Looking at `lif_address_unparse` only `inet_ntop`can fail and that has only two documented error cases which we both control. So I guess "this is fine".

File: cmd/ifparse.c [hunk 3]
```diff
@@ -47,24 +47,18 @@ set_allow_undefined(const char *arg)
 	allow_undefined = true;
 }
 
-#ifdef CONFIG_YAML
-static bool use_yaml = false;
+static const char *output_fmt = "ifupdown";
 
 static void
-set_yaml(const char *arg)
+set_output_fmt(const char *arg)
```
Comment (BarbarossaTM) [inline cmd/ifparse.c:53]: Shouldn't we `strcmp()` for `ifupdown` or `yaml` here?

File: cmd/ifparse.c [hunk 4]
```diff
@@ -47,24 +47,18 @@ set_allow_undefined(const char *arg)
...
 {
-	(void) arg;
-
-	use_yaml = true;
+	output_fmt = arg;
```
Comment (BarbarossaTM) [inline cmd/ifparse.c:55]: Shouldn't we `strcmp()` for `ifupdown` or `yaml` here?

Comment (kaniini) [reply cmd/ifparse.c:55]: nah it’s handled later by the lookup

**Pull request merged**